Add verification run scheduler

ByteOtter opened the pull request:
# What?

Add a script which wraps the standard use cases for `openqa-clone-custom-git-refspec` into a more usable form. Instead of requiring the user to type out the URLs of the `openqa.opensuse.org` test they want to use it injects all information - like the PR number, base test ID, test modules to load, etc. - into a `openqa-clone-custom-git-refspec` call.

This script itself does not implement new functionality or behaviour, but is simply a different "front-end" - with a more descriptive name - to use for quickly scheduling verification runs for PRs against e.g `os-autoinst`.

This is just a thought experiment and could theoretically be integrated into other components without the need for a separate script.

# How?

Several parameters are mandatory for the script to function:

- `$SOURCE` - Which specifies whether to use `openqa.opensuse.org` or `openqa.suse.de` (with custom targets potentially supported via config file)
- `$PR_ID` - The ID the GitHub PR
- `$BASE_TEST_ID` - The ID of the test the user wants to use as a base
- `$[test/module,...]` - A comma sperated list of test modules to load
- `$TEST_NAME` - The `TEST` parameter for the call

The GitHub access token is provided via a config file at `~/.config/gh.conf` (name temporary).

# Why?

This PR aims to lower the barrier of entry for new contributors and make general test writing easier by having a easy-to-use, descriptively named script a user can call to provide a run for their PR. Instead of going around trying to find the right command.

foursixnine commented:
Following the discussion slack: 

- This could be integrated inside [clone-custom-git-refspec](https://github.com/os-autoinst/openQA/blob/master/script/openqa-clone-custom-git-refspec) directly. That’s more due to how we already work, and I fear introducing a new script might not be 100% the best idea, but its a start already
- The handling of github tokens is a great adition!
- on the spec file, perhaps once we figure out what to do, it could be aliased to `openqa-verify-pr` or a nice short name :D (because `openqa-clone-custom-git-refspec` is... well, a bit long

kalikiana commented on script/openqa-verify_pr:
```diff
@@ -0,0 +1,126 @@
+elif [ "$source" = "suse" ]; then
+  echo "Dispatching verification run for suse..."
+  openqa-clonse-custom-git-refspec "https://github.com/os-autoinst-distri-opensuse/pull/$pr_id" \
+    "https://openqa.suse.de/tests/$test_id" \
```
I wonder, are you aware of the existing support for job URL's within the PR description? See https://github.com/os-autoinst/openQA/blob/master/script/openqa-clone-custom-git-refspec#L52

ByteOtter replied:
I actually wasn't aware of this, thanks for making me aware. That'd make integrating parts of the ideas into `openqa-clone-custom-git-refspec` even more alluring to take advantage of this feature. @foursixnine, what do you think?

foursixnine replied:
> I actually wasn't aware of this, thanks for making me aware. That'd make integrating parts of the ideas into `openqa-clone-custom-git-refspec` even more alluring to take advantage of this feature. @foursixnine, what do you think?

Yep its the first point here: https://github.com/os-autoinst/openQA/pull/5647#issuecomment-2117728546

Martchus commented:
Please capitalize the first letter of the commit message subject lines. It would also be much nicer if `openqa-verify_pr` was `openqa-verify-pr`. The formatting of the help text and general coding style is unfortunately not in-line with what we generally use in our scripts. I would at least add a `set -e` or perhaps even `set -eux -o pipefail` at the beginning.

Note that additional scripts could also be added to https://github.com/os-autoinst/scripts - although if it is supposed to be distributed via an RPM package keeping it here might be the better option.

ByteOtter commented:
Thanks for your feedback, I have changed the name and added the `set -o pipefail` line. Could you elaborate a bit on what you mean with "general coding style"? What do I need to do differently?
Thanks :)

Martchus commented on script/openqa-verify-pr:
```diff
@@ -0,0 +1,130 @@
+
+set -o pipefail
+
+source="$1"
+pr_id="$2"
+test_id="$3"
+modules="$4"
+test_name="$5"
+
+config_file="$HOME/.config/openqa/verification.conf"
+config=""
+custom_target=""
```
Unless shellcheck is enforcing them I'd get rid of these useless quotes and write everything in two lines:
```suggestion
source=$1 pr_id=$2 test_id=$3 modules=$4 test_name=$5
config_file=$HOME/.config/openqa/verification.conf config= custom_target=
```

ByteOtter replied:
These quotes were warnings of my LSP but your solution seems more compact. I will change it

Martchus replied:
Those quotes are really not required. The only annoyance can be that shellcheck will complain (needlessly). So I would check that first.

ByteOtter replied:
fixed it

Martchus commented on script/openqa-verify-pr:
```diff
@@ -0,0 +1,130 @@
+custom_target=""
+
+# See help text.
+if [ "$1" = "-h" ] || [ "$1" = "--help" ]; then
```
In Bash scripts it makes sense to use `[[` which also means that the left-hand-side doesn't need to be quoted. We usually also prefer single-quotes if no variable-substitution is used:
```suggestion
if [[ $1 = '-h' ] || [ $1 = '--help' ]; then
```

The same counts for the other occurrences as well.

Martchus commented on script/openqa-verify-pr:
```diff
@@ -0,0 +1,130 @@
+  exit 1
+fi
+
+GITHUB_TOKEN=$config
+
+export GITHUB_TOKEN
```
```suggestion
export GITHUB_TOKEN=$config
```

ByteOtter replied:
fixed it

Martchus commented on script/openqa-verify-pr:
```diff
@@ -0,0 +1,130 @@
+
+
+help_text=$(cat <<'EOF'
+OpenQA verification run scheduler.
```
```suggestion
openQA verification run scheduler
```

ByteOtter replied:
fixed it 👍🏼

Martchus commented on script/openqa-verify-pr:
```diff
@@ -0,0 +1,130 @@
+OpenQA verification run scheduler.
+
+Usage:
+ schedule_verification_run $SOURCE $PR_ID $BASE_TEST_ID $[test/module, test/module,...] $TEST_NAME
```
```suggestion
 openqa-verify-pr $SOURCE $PR_ID $BASE_TEST_ID $[test/module, test/module,...] $TEST_NAME
```

Martchus commented on script/openqa-verify-pr:
```diff
@@ -0,0 +1,130 @@
+# Schedule a verification run for a os-autoinst PR
+# USAGE:
+#
+# schedule_verification_run $SOURCE $PR_ID $BASE_TEST_ID $[test/module, test/module, ...] $TEST_NAME
```
```suggestion
# openqa-verify-pr $SOURCE $PR_ID $BASE_TEST_ID $[test/module, test/module, ...] $TEST_NAME
```

But considering this is redundant to the help text I would just remove this lengthy comment-block completely.

ByteOtter replied:
fixed it 👍🏼

Martchus commented on script/openqa-verify-pr:
```diff
@@ -0,0 +1,130 @@
+#!/bin/bash
+# Copyright 2024 (c) SUSE LLC
+# SPDX-License-Identifier: GPL-2.0-or-later
```
```suggestion
# Copyright SUSE LLC
# SPDX-License-Identifier: GPL-2.0-or-later
```

ByteOtter replied:
fixed 👍🏼

Martchus commented on script/openqa-verify-pr:
```diff
@@ -0,0 +1,130 @@
+
+Parameters:
+ SOURCE - The source you want to schedule a run for. Can be either opensuse or suse.
+ PR_ID - The ID of your GitHub Pull Request.
```
```suggestion
 PR_ID - The ID of your GitHub pull request.
```
(in accordance with https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-pull-requests)

ByteOtter replied:
fixed 👍🏼

Martchus commented on script/openqa-verify-pr:
```diff
@@ -0,0 +1,130 @@
+ schedule_verification_run $SOURCE $PR_ID $BASE_TEST_ID $[test/module, test/module,...] $TEST_NAME
+
+Parameters:
+ SOURCE - The source you want to schedule a run for. Can be either opensuse or suse.
```
```suggestion
 SOURCE - The source you want to schedule a run for. Can be either openSUSE or SUSE.
```

ByteOtter replied:
In this case I wanted to print the actual values you can pass to the script. but it's outdated anyway so I'll change it

ByteOtter replied:
Marked `opensuse` and `suse` with single quotes to underline them being values to pass rather than simple name listing

Martchus commented on script/openqa-verify-pr:
```diff
@@ -0,0 +1,130 @@
+
+CUSTOM_TARGET="https://openqa.yourdomain.com"
+
+Default test modules:
+The following modules are automatically selected to load for your run:
```
Two consecutive phrases ending with a colon are a bit weird.

ByteOtter replied:
Leftover from what I had `Default test modules` as a heading for that section. I'll change that.

Martchus commented on script/openqa-verify-pr:
```diff
@@ -0,0 +1,130 @@
+config=""
+custom_target=""
+
+# See help text.
```
```suggestion
```

Martchus commented on script/openqa-verify-pr:
```diff
@@ -0,0 +1,130 @@
+  exit 0
+fi
+
+# If too few or too many arguments are supplied -> die.
```
```suggestion
# Validate argument count
```

Martchus reviewed (commented):
Generally we use a more compact coding style. I gave a few concrete suggestions in inline-comments.

foursixnine commented on script/openqa-clone-custom-git-refspec:
```diff
@@ -32,6 +32,9 @@ set -o pipefail
 if [[ -n "$GITHUB_TOKEN" ]]; then
     AUTHENTICATED_REQUEST=" -u $GITHUB_TOKEN:x-oauth-basic"
     echo "Github oauth token provided, performing authenticated requests"
+elif [[ -f "$HOME/.config/openqa/verification.conf" ]]; then
```
```suggestion
elif [[ -f "$HOME/.config/openqa/github-token.conf" ]]; then
```

foursixnine commented on script/openqa-clone-custom-git-refspec:
```diff
@@ -55,6 +58,16 @@ extract_urls_from_pr() {
     fi
 }
 
+# Get the Github Token from a config file
+# located at ~/.config/openqa/verification.conf
+get_token_from_file() {
+  local config_file="$HOME/.config/openqa/verification.conf"
```
See above, also provide an example of the format in the help file.

ByteOtter replied:
fixed 👍🏼

foursixnine commented on script/openqa-verify-pr:
```diff
@@ -0,0 +1,119 @@
+  echo "Dispatching verification run for opensuse..."
+  openqa-clone-custom-git-refspec "https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/$pr_id" \
+    "https://openqa.opensuse.org/tests/$test_id" \
+    SCHEDULE="tests/installation/bootloader_start,tests/boot/boot_to_desktop,$modules" \
```
Remember to move this to the  [openqa-clone-custom-git-refspec](https://github.com/os-autoinst/openQA/pull/5647/files#diff-1da53224de38473695839b4d4c7efee7a2add51a1aa9a3c259e872a5663b9498) behind a switch i.e `--modules a,b,c` so it gets appended via the for loop, before calling cone job.

and then there's the consideration that if the caller is passing:
- YAML_SCHEDULE + SCHEDULE, die (as one of them has to be overwritten.
- if the switch is passed, and so is either (yaml_schedule or schedule) die, as options aren't compatible

ByteOtter replied:
Yep, currently looking into that :)

foursixnine commented on script/openqa-clone-custom-git-refspec:
```diff
@@ -157,6 +182,12 @@ Please try 'curl $json_url' or select another job, e.g. in the same scenario: $h
     local scriptdir
     scriptdir=$(dirname "${BASH_SOURCE[0]}")
     local cmd="$dry_run $scriptdir/openqa-clone-job $clone_args \"$host\" \"$job\" _GROUP=\"$GROUP\" TEST+=\"$test_suffix\" BUILD=\"$build\" CASEDIR=\"$casedir\" PRODUCTDIR=\"$productdir\" NEEDLES_DIR=\"$needles_dir\""
+    if [[ $cmd =~ (SCHEDULE|YAML_SCHEDULE) ]]; then
```
is this an and? or an or?

ByteOtter replied:
it's an or, my bad. Though changing it to
```bash
if [[ $cmd =~ SCHEDULE ]] && [[ $cmd =~ YAML_SCHEDULE ]]; then
...
```
doesnt seem to change anythin rn so my problem may be that I am using `cmd`?

okurz reviewed (changes requested):
As suggested in chat originally I appreciate that you try to bring in new ideas with pull requests but we should be careful about introducing new scripts. And the same probably goes for the modules command line parameter which is redundant considering the support for generic openQA test variables. Before investing more effort into that I recommend you look into this variables concept in general. Maybe for now you want to focus only on the aspect of reading the github token from a config file which seems independent of the other changes

ByteOtter commented on script/openqa-clone-custom-git-refspec:
```diff
@@ -12,6 +12,7 @@ Options:
  -v, --verbose          execute with verbose output
  -h, -?, --help         display this help
  -n, --dry-run          execute in dry-run mode, do not clone any openQA jobs
+ -m, --modules          specify the tets modules you want to load as a comma-separated list of paths to your test files
```
To separate the concerns of the `--modules` flag and the token handling, I extracted the token handling into its own PR https://github.com/os-autoinst/openQA/pull/5666 for now.

foursixnine replied:
This is added as a comodity command, the schedule, include_modules and exclude modules are like a shotgun: You can blow your foot if you want to.

`--modules` is more like a hammer, you can still smash your fingers, but won't have you hunting for reasons why your forced schedule is not being taken into account. See line 200.
